src/main/java/build/buildfarm/worker/CASFileCache.java

```diff
@@ -272,21 +279,33 @@ private void fetchDirectory(
     }
   }
 
-  public synchronized Path putDirectory(
+  public Path putDirectory(
       Digest digest,
       Map<Digest, Directory> directoriesIndex) throws IOException, InterruptedException {
     Path path = getDirectoryPath(digest);
 
-    DirectoryEntry e = directoryStorage.get(digest);
-    if (e != null) {
-      incrementReferences(e.inputs);
-      return path;
+    synchronized (this) {
+      DirectoryEntry e = directoryStorage.get(digest);
+      if (e != null) {
+        incrementReferences(e.inputs);
+        return path;
+      }
     }
 
     ImmutableList.Builder<Path> inputsBuilder = new ImmutableList.Builder<>();
-    fetchDirectory(digest, path, digest, directoriesIndex, inputsBuilder);
+    try {
+      fetchDirectory(digest, path, digest, directoriesIndex, inputsBuilder);
+    } catch (IOException e) {
```

Comment:
What if there's another IOException while deleting the directory? You probably don't want it to hide the original exception.

Reply:
Agreed. Added a catch with a printStackTrace and rerouted.